geomstats/stiefel.py

```diff
@@ -66,3 +70,289 @@ def random_uniform(self, n_samples=1):
         inv_sqrt_aux = gs.linalg.inv(sqrt_aux)
 
         return gs.matmul(std_normal, inv_sqrt_aux)
+
+
+class StiefelCanonicalMetric(RiemannianMetric):
+
+    def __init__(self, n, p):
+        dimension = int(p * n - (p * (p + 1) / 2))
+        super(StiefelCanonicalMetric, self).__init__(
+                dimension=dimension,
+                signature=(dimension, 0, 0))
+        self.embedding_metric = EuclideanMetric(n*p)
+        self.n = n
+        self.p = p
+
+    def inner_product(self, tangent_vec_a, tangent_vec_b, base_point):
+        """
+        Canonical inner product on the tangent space at base_point,
+        which is different from the inner product induced by the embedding.
+
+        Formula from:
+        http://noodle.med.yale.edu/hdtag/notes/steifel_notes.pdf
+        """
+        tangent_vec_a = gs.to_ndarray(tangent_vec_a, to_ndim=3)
+        tangent_vec_b = gs.to_ndarray(tangent_vec_b, to_ndim=3)
+        base_point = gs.to_ndarray(base_point, to_ndim=3)
+        base_point_transpose = gs.transpose(base_point, axes=(0, 2, 1))
+
+        aux = gs.matmul(
+            gs.transpose(tangent_vec_a, axes=(0, 2, 1)),
+            gs.eye(self.n) - 0.5 * gs.matmul(base_point, base_point_transpose))
+        inner_prod = gs.trace(gs.matmul(aux, tangent_vec_b), axis1=1, axis2=2)
+
+        inner_prod = gs.to_ndarray(inner_prod, to_ndim=2, axis=1)
+        return inner_prod
+
+    def exp(self, tangent_vec, base_point):
+        """
+        Riemannian exponential of a tangent vector wrt to a base point.
+        """
+        tangent_vec = gs.to_ndarray(tangent_vec, to_ndim=3)
```

Comment:
note for later: we should really be able to refactor all these functions somehow, with a decorator or something similar.

not in this PR, but let's discuss in person

Reply:
Completely agree.